### PR TITLE
Fix v3 resource deduplication.

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Since the v3 resources have an embedded resource version, dedupe was always
failing.

May fix https://github.com/projectcalico/calico/issues/4129

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by fasaxc** [comment]
Wrote a UT locally but it doesn't pass, needs more work....

**Comment by fasaxc** [comment]
Fixed the UT, good to go now...

File: pkg/syncproto/sync_proto.go [hunk 1]
```diff
@@ -259,6 +260,16 @@ func SerializeUpdate(u api.Update) (su SerializedUpdate, err error) {
 	if u.Value == nil {
 		log.Debug("Value is nil, passing through as a deletion.")
 		return
+	} else if obj, ok := u.Value.(v1.Object); ok {
+		// Since v3 objectsgit commi carry their resource version inside their internal metadata, our
```
[the remark this answers is not in the document]
Comment (neiljerram) [reply pkg/syncproto/sync_proto.go:264]: Also maybe worth saying "Calico v3 objects ..."

File: pkg/syncproto/sync_proto.go [hunk 2]
```diff
@@ -259,6 +260,16 @@ func SerializeUpdate(u api.Update) (su SerializedUpdate, err error) {
...
+		// later dedupe comparison will always fail unless we zero this out. The KVP revision
+		// should always be the same as the internal revision but we copy it over just to make
+		// sure that we replace the revision correctly on the client side.
```
Comment (neiljerram) [inline pkg/syncproto/sync_proto.go:267]: Why not use a distinct new SerializedUpdate field to hold the internal metadata revision?

**Review by neiljerram — CHANGES_REQUESTED** [review]
A few queries and typos

File: pkg/syncproto/sync_proto.go (hunk 2 above)
Comment (fasaxc) [reply pkg/syncproto/sync_proto.go:267]: Yeah, that might work well.  Good thought.

File: pkg/syncproto/sync_proto.go [hunk 3]
```diff
@@ -259,6 +260,16 @@ func SerializeUpdate(u api.Update) (su SerializedUpdate, err error) {
 	if u.Value == nil {
 		log.Debug("Value is nil, passing through as a deletion.")
 		return
+	} else if obj, ok := u.Value.(v1.Object); ok {
+		// Since v3 objectsgit commi carry their resource version inside their internal metadata, our
+		// later dedupe comparison will always fail unless we zero this out. The KVP revision
+		// should always be the same as the internal revision but we copy it over just to make
```
[the remark this answers is not in the document]
Comment (fasaxc) [reply pkg/syncproto/sync_proto.go]: The revision in the key-value pair struct and the one that's embedded inside the resource.

**Review by neiljerram — APPROVED** [review]
LGTM

**Pull request merged**